packages/contracts/src/contracts/current/forwarder/MixinForwarderQuote.sol

```diff
@@ -0,0 +1,139 @@
+pragma solidity ^0.4.21;
+pragma experimental ABIEncoderV2;
+
+import "./MixinForwarderCore.sol";
+
+contract MixinForwarderQuote is MixinForwarderCore {
+
+    /// @dev Performs the 0x Exchange validation and calculations, without performing any state changes.
+    /// @param order An Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function fillOrderQuote(Order memory order, uint256 takerAssetFillAmount)
+        public
+        view
+        returns (Exchange.FillResults memory fillResults)
+    {
+        // Compute the order hash
+        bytes32 orderHash = getOrderHash(order);
+        uint256 remainingTakerAssetAmount = safeSub(order.takerAssetAmount, EXCHANGE.getUnavailableTakerTokenAmount(orderHash));
+        // Validate order expiration
+        if (block.timestamp >= order.expirationTimeSeconds) {
+            return fillResults;
+        }
+        // Validate order availability
+        fillResults.takerAssetFilledAmount = min256(takerAssetFillAmount, remainingTakerAssetAmount);
+        if (fillResults.takerAssetFilledAmount == 0) {
+            return fillResults;
+        }
+        // Validate fill order rounding
+        if (isRoundingError(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount)) {
+            fillResults.takerAssetFilledAmount = 0;
+            return fillResults;
+        }
+        fillResults.makerAssetFilledAmount = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount);
+        if (order.feeRecipientAddress != address(0)) {
```

Comment:
This is out-of-date with V2. We no longer enforce that `feeRecipientAddress` is not NULL_ADDRESS. I guess I'm concerned with us keeping this logic up-to-date with Exchange in general but don't see a way around this atm.